File: implementations/python/core/advisor.py
"""FIRE Advisor - Provides optimization suggestions based on calculation results."""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

import pandas as pd

from .data_models import SimulationSettings
from .engine import EngineInput, FIREEngine
from .monte_carlo import MonteCarloSimulator


@dataclass
class SimpleRecommendation:
    """Simple recommendation with type and parameters for UI rendering."""

    type: str
    params: Dict[str, Any]
    is_achievable: bool = True
    monte_carlo_success_rate: Optional[float] = None


class FIREAdvisor:
    """FIRE Advisor that provides structured optimization recommendations.

    This advisor generates language-agnostic structured data that can be
    dynamically rendered in any language by the UI layer.
    """

    def __init__(self, engine_input: EngineInput, language: Optional[str] = None):
        """Initialize advisor with engine input.

        Args:
            engine_input: Engine input with user profile and projections
            language: Deprecated parameter, kept for backward compatibility
        """
        self.engine_input = engine_input
        self.profile = engine_input.user_profile
        self.projection_df = engine_input.annual_financial_projection
        self.detailed_projection_df = engine_input.detailed_projection
        self.income_items = engine_input.income_items

    def get_all_recommendations(self) -> List[SimpleRecommendation]:
        """Get all advisor recommendations based on current FIRE feasibility.

        Returns:
            List of SimpleRecommendation objects with type and parameters
        """
        # First check if current plan is achievable
        engine = FIREEngine(self.engine_input)
        base_result = engine.calculate()

        recommendations: List[SimpleRecommendation] = []

        if base_result.is_fire_achievable:
            # Plan is achievable - find earliest possible retirement
            early_rec = self._find_earliest_retirement()
            if early_rec:
                recommendations.append(early_rec)
        else:
            # Plan not achievable - provide multiple solutions
            delayed_rec = self._find_required_delayed_retirement()
            income_rec = self._find_required_income_increase()
            expense_rec = self._find_required_expense_reduction()

            for rec in [delayed_rec, income_rec, expense_rec]:
                if rec:
                    recommendations.append(rec)

        return recommendations

    def _find_earliest_retirement(self) -> Optional[SimpleRecommendation]:
        """Find the earliest possible retirement age."""
        current_expected_age = self.profile.expected_fire_age
        current_age = self.profile.current_age

        # Start from one year earlier and work backwards
        test_age = current_expected_age - 1
        earliest_achievable_age = current_expected_age

        while test_age >= current_age:
            # Create modified profile with earlier FIRE age
            modified_profile = self.profile.model_copy()
            modified_profile.expected_fire_age = test_age

            # Create modified projection with truncated work income
            modified_detailed_projection = self._truncate_work_income_to_age(test_age)
            modified_annual_projection = self._create_annual_summary_from_detailed_df(
                modified_detailed_projection
            )

            # Test this age
            modified_input = EngineInput(
                user_profile=modified_profile,
                annual_financial_projection=modified_annual_projection,
                detailed_projection=modified_detailed_projection,
                income_items=self.income_items,
            )

            engine = FIREEngine(modified_input)
            result = engine.calculate()

            if result.is_fire_achievable:
                earliest_achievable_age = test_age
                test_age -= 1
            else:
                break

        if earliest_achievable_age < current_expected_age:
            # Get calculation results for the optimal age
            optimal_profile = self.profile.model_copy()
            optimal_profile.expected_fire_age = earliest_achievable_age

            # Create projection with truncated income for optimal age
            optimal_detailed_projection = self._truncate_work_income_to_age(
                earliest_achievable_age
            )
            optimal_annual_projection = self._create_annual_summary_from_detailed_df(
                optimal_detailed_projection
            )
            optimal_input = EngineInput(
                user_profile=optimal_profile,
                annual_financial_projection=optimal_annual_projection,
                detailed_projection=optimal_detailed_projection,
                income_items=self.income_items,
            )

            engine = FIREEngine(optimal_input)
            engine.calculate()

            # Run Monte Carlo for this optimal age
            optimal_engine = FIREEngine(optimal_input)
            mc_simulator = MonteCarloSimulator(
                optimal_engine,
                SimulationSettings(
                    num_simulations=1000,
                    confidence_level=0.95,
                    include_black_swan_events=True,
                    income_base_volatility=0.1,
                    income_minimum_factor=0.1,
                    expense_base_volatility=0.05,
                    expense_minimum_factor=0.5,
                ),
            )
            mc_result = mc_simulator.run_simulation()

            years_saved = current_expected_age - earliest_achievable_age

            return SimpleRecommendation(
                type="early_retirement",
                params={"age": earliest_achievable_age, "years": years_saved},
                is_achievable=True,
                monte_carlo_success_rate=mc_result.success_rate,
            )

        return None

    def _find_required_delayed_retirement(
        self,
    ) -> Optional[SimpleRecommendation]:
        """Find the minimum age required for FIRE to be achievable."""
        current_expected_age = self.profile.expected_fire_age
        legal_retirement_age = self.profile.legal_retirement_age

        # Start from one year later and work forward
        test_age = current_expected_age + 1
        required_age = None

        while test_age <= legal_retirement_age:
            # Create modified profile with later FIRE age
            modified_profile = self.profile.model_copy()
            modified_profile.expected_fire_age = test_age

            # Create modified detailed projection with extended work income
            modified_detailed_projection = self._extend_work_income_to_age(test_age)

            # Convert to annual summary for engine
            modified_annual_projection = self._create_annual_summary_from_detailed_df(
                modified_detailed_projection
            )

            # Test this age
            modified_input = EngineInput(
                user_profile=modified_profile,
                annual_financial_projection=modified_annual_projection,
                detailed_projection=modified_detailed_projection,
                income_items=self.income_items,
            )

            engine = FIREEngine(modified_input)
            result = engine.calculate()

            if result.is_fire_achievable:
                required_age = test_age
                break

            test_age += 1

        if required_age:
            # Get calculation results for the required age
            required_profile = self.profile.model_copy()
            required_profile.expected_fire_age = required_age
            required_detailed_projection = self._extend_work_income_to_age(required_age)
            required_annual_projection = self._create_annual_summary_from_detailed_df(
                required_detailed_projection
            )
            required_input = EngineInput(
                user_profile=required_profile,
                annual_financial_projection=required_annual_projection,
                detailed_projection=required_detailed_projection,
                income_items=self.income_items,
            )

            engine = FIREEngine(required_input)
            engine.calculate()

            years_delayed = required_age - current_expected_age

            return SimpleRecommendation(
                type="delayed_retirement",
                params={"age": required_age, "years": years_delayed},
                is_achievable=True,
            )

        # If no feasible age found within legal retirement age, return unfeasible
        # recommendation
        # Test the maximum possible delay (to legal retirement age)
        max_delay_age = legal_retirement_age
        max_delay_profile = self.profile.model_copy()
        max_delay_profile.expected_fire_age = max_delay_age

        max_delay_detailed_projection = self._extend_work_income_to_age(max_delay_age)
        max_delay_annual_projection = self._create_annual_summary_from_detailed_df(
            max_delay_detailed_projection
        )

        max_delay_input = EngineInput(
            user_profile=max_delay_profile,
            annual_financial_projection=max_delay_annual_projection,
            detailed_projection=max_delay_detailed_projection,
            income_items=self.income_items,
        )

        engine = FIREEngine(max_delay_input)
        engine.calculate()

        years_delayed = max_delay_age - current_expected_age

        return SimpleRecommendation(
            type="delayed_retirement_not_feasible",
            params={"age": max_delay_age},
            is_achievable=False,
        )

    def _find_required_income_increase(
        self,
    ) -> Optional[SimpleRecommendation]:
        """Find required income multiplier to achieve FIRE at expected age."""
        # Binary search for the minimum income multiplier
        low, high = 1.0, 5.0  # Search between 1x and 5x income
        epsilon = 0.01  # Precision for binary search

        optimal_multiplier = None

        while high - low > epsilon:
            mid = (low + high) / 2

            # Create modified projection with increased income
            modified_df = self.projection_df.copy()
            modified_df["total_income"] *= mid

            modified_input = EngineInput(
                user_profile=self.profile, annual_financial_projection=modified_df
            )

            engine = FIREEngine(modified_input)
            result = engine.calculate()

            if result.is_fire_achievable:
                optimal_multiplier = mid
                high = mid
            else:
                low = mid

        if optimal_multiplier:
            # Calculate additional income needed
            original_income = self.projection_df["total_income"].iloc[0]
            additional_income = original_income * (optimal_multiplier - 1.0)

            # Get final calculation result
            final_df = self.projection_df.copy()
            final_df["total_income"] *= optimal_multiplier
            final_input = EngineInput(
                user_profile=self.profile, annual_financial_projection=final_df
            )

            engine = FIREEngine(final_input)
            engine.calculate()

            return SimpleRecommendation(
                type="increase_income",
                params={
                    "fire_age": self.profile.expected_fire_age,
                    "percentage": (optimal_multiplier - 1) * 100,
                    "amount": additional_income,
                },
                is_achievable=True,
            )

        return None

    def _find_required_expense_reduction(
        self,
    ) -> Optional[SimpleRecommendation]:
        """Find required expense reduction to achieve FIRE at expected age."""
        # Binary search for the minimum expense reduction
        low, high = 0.0, 0.8  # Search between 0% and 80% reduction
        epsilon = 0.001  # Precision for binary search

        optimal_reduction = None

        while high - low > epsilon:
            mid = (low + high) / 2
            reduction_factor = 1.0 - mid  # Convert reduction rate to multiplier

            # Create modified projection with reduced expenses
            modified_df = self.projection_df.copy()
            modified_df["total_expense"] *= reduction_factor

            modified_input = EngineInput(
                user_profile=self.profile, annual_financial_projection=modified_df
            )

            engine = FIREEngine(modified_input)
            result = engine.calculate()

            if result.is_fire_achievable:
                optimal_reduction = mid
                high = mid
            else:
                low = mid

        if optimal_reduction:
            # Calculate annual savings needed
            original_expense = self.projection_df["total_expense"].iloc[0]
            annual_savings = original_expense * optimal_reduction

            # Get final calculation result
            final_df = self.projection_df.copy()
            final_df["total_expense"] *= 1.0 - optimal_reduction
            final_input = EngineInput(
                user_profile=self.profile, annual_financial_projection=final_df
            )

            engine = FIREEngine(final_input)
            engine.calculate()

            return SimpleRecommendation(
                type="reduce_expenses",
                params={
                    "fire_age": self.profile.expected_fire_age,
                    "percentage": optimal_reduction * 100,
                    "amount": annual_savings,
                },
                is_achievable=True,
            )

        return None

    def _extend_work_income_to_age(self, target_fire_age: int) -> pd.DataFrame:
        """Create modified detailed projection with extended work income.

        This method identifies work income items (typically those that end at or before
        current expected_fire_age) and extends them to the new target age.

        Args:
            target_fire_age: New target FIRE age to extend work income to

        Returns:
            Modified detailed projection DataFrame with extended work income
        """
        if self.detailed_projection_df is None or not self.income_items:
            raise ValueError("Detailed projection data required for income extension")

        extended_projection = self.detailed_projection_df.copy()
        current_fire_age = self.profile.expected_fire_age

        # If target age is not later than current, return unchanged
        if target_fire_age <= current_fire_age:
            return extended_projection

        # Find income items that end at current FIRE age (work income)
        work_income_items = [
            item
            for item in self.income_items
            if item.end_age and item.end_age == current_fire_age
        ]

        if not work_income_items:
            # No work income to extend
            return extended_projection

        # For each work income item, extend it to target age
        for item in work_income_items:
            # Calculate the growth pattern for extending this income
            for age in range(item.end_age + 1, target_fire_age + 1):
                age_mask = extended_projection["age"] == age
                if age_mask.any():
                    row_idx = extended_projection[age_mask].index[0]

                    # Calculate extended income value with proper growth
                    years_since_start = age - item.start_age
                    growth_factor = (
                        1 + item.annual_growth_rate / 100
                    ) ** years_since_start
                    extended_income = item.after_tax_amount_per_period * growth_factor

                    extended_projection.loc[row_idx, item.name] = extended_income

        return extended_projection

    def _create_annual_summary_from_detailed_df(
        self, detailed_df: pd.DataFrame
    ) -> pd.DataFrame:
        """Create annual summary from detailed projection DataFrame.

        This converts the wide format detailed projection (with individual
        income/expense columns) to the summary format expected by FIREEngine.
        """
        if not self.income_items:
            raise ValueError("Income items required for summary creation")

        # Get income column names
        income_cols = [item.name for item in self.income_items]

        # Get expense columns (all columns except 'age', 'year', and income columns)
        all_cols = set(detailed_df.columns)
        system_cols = {"age", "year"}
        income_cols_set = set(income_cols)
        expense_cols = list(all_cols - system_cols - income_cols_set)

        # Calculate totals
        result_df = detailed_df[["age", "year"]].copy()
        result_df["total_income"] = detailed_df[income_cols].sum(axis=1)
        result_df["total_expense"] = (
            detailed_df[expense_cols].sum(axis=1) if expense_cols else 0.0
        )
        result_df["net_flow"] = result_df["total_income"] - result_df["total_expense"]

        return result_df

    def _truncate_work_income_to_age(self, target_fire_age: int) -> pd.DataFrame:
        """Create modified detailed projection with truncated work income.

        This method identifies work income items (typically those that end at or after
        current expected_fire_age) and truncates them at the new target age.

        Args:
            target_fire_age: New target FIRE age to truncate work income at

        Returns:
            Modified detailed projection DataFrame with work income truncated at
            target age
        """
        if self.detailed_projection_df is None or not self.income_items:
            raise ValueError("Detailed projection data required for income truncation")

        truncated_projection = self.detailed_projection_df.copy()
        current_fire_age = self.profile.expected_fire_age

        # If target age is not earlier than current, return unchanged
        if target_fire_age >= current_fire_age:
            return truncated_projection

        # Find income items that end at current FIRE age (work income)
        work_income_items = [
            item
            for item in self.income_items
            if item.end_age and item.end_age == current_fire_age
        ]

        if not work_income_items:
            # No work income to truncate
            return truncated_projection

        # For each work income item, truncate it at target retirement age
        for item in work_income_items:
            # Set income to 0 for ages after target retirement age
            for age in range(target_fire_age + 1, item.end_age + 1):
                age_mask = truncated_projection["age"] == age
                if age_mask.any():
                    truncated_projection.loc[age_mask, item.name] = 0

        return truncated_projection

File: implementations/typescript/src/core/advisor.ts
/**
 * FIRE Advisor - Provides optimization suggestions based on calculation results
 * Direct TypeScript port from Python advisor.py
 *
 * This module analyzes FIRE plans and provides structured recommendations for:
 * - Early retirement opportunities
 * - Delayed retirement solutions
 * - Income increase requirements
 * - Expense reduction strategies
 */

import { Decimal } from 'decimal.js';
import type {
  UserProfile,
  SimulationSettings,
  IncomeExpenseItem,
} from './data_models';
import { createSimulationSettings, getCurrentAgeAsOf } from './data_models';
import type {
  EngineInput,
  AnnualFinancialProjection,
  DetailedProjection,
} from './engine';
import { FIREEngine, createEngineInput } from './engine';
import { MonteCarloSimulator } from './monte_carlo';

// =============================================================================
// Data Models
// =============================================================================

/**
 * Simple recommendation with type and parameters for UI rendering
 * Direct TypeScript equivalent of Python's SimpleRecommendation dataclass
 */
export interface SimpleRecommendation {
  /** Recommendation type (early_retirement, delayed_retirement, etc.) */
  type: string;

  /** Parameters for UI rendering */
  params: Record<string, any>;

  /** Whether this recommendation makes FIRE achievable */
  is_achievable: boolean;

  /** Monte Carlo success rate for this recommendation */
  monte_carlo_success_rate?: Decimal;
}

// =============================================================================
// FIRE Advisor - Main Class
// =============================================================================

/**
 * FIRE Advisor that provides structured optimization recommendations
 * Direct TypeScript equivalent of Python's FIREAdvisor class
 *
 * This advisor generates language-agnostic structured data that can be
 * dynamically rendered in any language by the UI layer.
 */
export class FIREAdvisor {
  public readonly engine_input: EngineInput;
  public readonly profile: UserProfile;
  public readonly projection_df: AnnualFinancialProjection[];
  public readonly detailed_projection_df:
    | AnnualFinancialProjection[]
    | DetailedProjection[];
  public readonly income_items: IncomeExpenseItem[];

  constructor(engine_input: EngineInput, language?: string) {
    // language parameter is deprecated but kept for backward compatibility
    this.engine_input = engine_input;
    this.profile = engine_input.user_profile;
    this.projection_df = engine_input.annual_financial_projection;
    this.detailed_projection_df = [...engine_input.annual_financial_projection]; // Create a copy
    this.income_items = engine_input.income_items || [];
  }

  /**
   * Get engine input - compatibility method for tests
   */
  getEngineInput(): EngineInput {
    return this.engine_input;
  }

  /**
   * Get all advisor recommendations - compatibility method for tests
   * Delegates to the Python-style method name
   */
  async getAllRecommendations(): Promise<SimpleRecommendation[]> {
    return await this.get_all_recommendations();
  }

  /**
   * Get all advisor recommendations based on current FIRE feasibility
   */
  async get_all_recommendations(): Promise<SimpleRecommendation[]> {
    // First check if current plan is achievable
    const engine = new FIREEngine(this.engine_input);
    const base_result = engine.calculate();

    const recommendations: SimpleRecommendation[] = [];

    if (base_result.is_fire_achievable) {
      // Plan is achievable - find earliest possible retirement
      const early_rec = await this._find_earliest_retirement();
      if (early_rec) {
        recommendations.push(early_rec);
      }
    } else {
      // Plan not achievable - provide multiple solutions
      const delayed_rec = this._find_required_delayed_retirement();
      const income_rec = this._find_required_income_increase();
      const expense_rec = this._find_required_expense_reduction();

      // Filter out null recommendations (matching Python behavior)
      for (const rec of [delayed_rec, income_rec, expense_rec]) {
        if (rec) {
          recommendations.push(rec);
        }
      }
    }

    return recommendations;
  }

  /**
   * Find the earliest possible retirement age using binary search
   */
  private async _find_earliest_retirement(): Promise<SimpleRecommendation | null> {
    const current_expected_age = this.profile.expected_fire_age;
    const current_age = getCurrentAgeAsOf(
      this.profile.birth_year,
      this.profile.as_of_year ?? new Date().getFullYear()
    );

    // Start from one year earlier and work backwards
    let test_age = current_expected_age - 1;
    let earliest_achievable_age = current_expected_age;

    while (test_age >= current_age) {
      // Create modified profile with earlier FIRE age
      const modified_profile: UserProfile = {
        ...this.profile,
        expected_fire_age: test_age,
      };

      // Create modified projection with truncated work income
      const modified_detailed_projection =
        this._truncate_work_income_to_age(test_age);
      const modified_annual_projection =
        this._create_annual_summary_from_detailed_df(
          modified_detailed_projection
        );

      // Test this age
      const modified_input = createEngineInput(
        modified_profile,
        modified_annual_projection,
        this.income_items
      );

      const engine = new FIREEngine(modified_input);
      const result = engine.calculate();

      if (result.is_fire_achievable) {
        earliest_achievable_age = test_age;
        test_age -= 1;
      } else {
        break;
      }
    }

    if (earliest_achievable_age < current_expected_age) {
      // Get calculation results for the optimal age
      const optimal_profile: UserProfile = {
        ...this.profile,
        expected_fire_age: earliest_achievable_age,
      };

      // Create projection with truncated income for optimal age
      const optimal_detailed_projection = this._truncate_work_income_to_age(
        earliest_achievable_age
      );
      const optimal_annual_projection =
        this._create_annual_summary_from_detailed_df(
          optimal_detailed_projection
        );
      const optimal_input = createEngineInput(
        optimal_profile,
        optimal_annual_projection,
        this.income_items
      );

      const engine = new FIREEngine(optimal_input);
      engine.calculate();

      // Run Monte Carlo for this optimal age
      const optimal_engine = new FIREEngine(optimal_input);
      const mc_simulator = new MonteCarloSimulator(
        optimal_engine,
        createSimulationSettings({
          num_simulations: 1000,
          confidence_level: new Decimal(0.95),
          include_black_swan_events: true,
          income_base_volatility: new Decimal(0.1),
          income_minimum_factor: new Decimal(0.1),
          expense_base_volatility: new Decimal(0.05),
          expense_minimum_factor: new Decimal(0.5),
        })
      );
      const mc_result = await mc_simulator.run_simulation();

      const years_saved = current_expected_age - earliest_achievable_age;

      return {
        type: 'early_retirement',
        params: {
          age: earliest_achievable_age,
          years: years_saved,
          suggested_fire_age: earliest_achievable_age,
        },
        is_achievable: true,
        monte_carlo_success_rate: mc_result.success_rate,
      };
    }

    return null;
  }

  /**
   * Find the minimum age required for FIRE to be achievable
   */
  private _find_required_delayed_retirement(): SimpleRecommendation | null {
    const current_expected_age = this.profile.expected_fire_age;
    const legal_retirement_age = this.profile.legal_retirement_age;

    // Start from one year later and work forward
    let test_age = current_expected_age + 1;
    let required_age: number | null = null;

    while (test_age <= legal_retirement_age) {
      // Create modified profile with later FIRE age
      const modified_profile: UserProfile = {
        ...this.profile,
        expected_fire_age: test_age,
      };

      // Create modified detailed projection with extended work income
      const modified_detailed_projection =
        this._extend_work_income_to_age(test_age);

      // Convert to annual summary for engine
      const modified_annual_projection =
        this._create_annual_summary_from_detailed_df(
          modified_detailed_projection
        );

      // Test this age
      const modified_input = createEngineInput(
        modified_profile,
        modified_annual_projection,
        this.income_items
      );

      const engine = new FIREEngine(modified_input);
      const result = engine.calculate();

      if (result.is_fire_achievable) {
        required_age = test_age;
        break;
      }

      test_age += 1;
    }

    if (required_age) {
      const years_delayed = required_age - current_expected_age;

      return {
        type: 'delayed_retirement',
        params: {
          age: required_age,
          years: years_delayed,
        },
        is_achievable: true,
        monte_carlo_success_rate: undefined, // Match Python's null
      };
    }

    // If no feasible solution found within legal retirement age, still recommend delaying to legal age
    // This matches Python behavior of always trying to provide a solution
    const years_delayed = legal_retirement_age - current_expected_age;

    return {
      type: 'delayed_retirement_not_feasible',
      params: {
        age: legal_retirement_age,
      },
      is_achievable: false,
      monte_carlo_success_rate: undefined,
    };
  }

  /**
   * Find required income increase using binary search
   */
  private _find_required_income_increase(): SimpleRecommendation | null {
    const max_multiplier = new Decimal(5.0); // Maximum 500% income increase
    const precision = new Decimal(0.01); // Match Python's epsilon = 0.01
    let low = new Decimal(1.0);
    let high = max_multiplier;
    let required_multiplier: Decimal | null = null;

    let iteration = 0;
    // Binary search for minimum income multiplier
    while (high.sub(low).gt(precision)) {
      const mid = low.add(high).div(2);
      iteration++;

      // Create modified projection with increased income
      const modified_projection = this._apply_income_multiplier(mid);

      // Test this multiplier
      const modified_input = createEngineInput(
        this.profile,
        modified_projection,
        this.income_items
      );

      const engine = new FIREEngine(modified_input);
      const result = engine.calculate();

      if (result.is_fire_achievable) {
        required_multiplier = mid;
        high = mid;
      } else {
        low = mid;
      }

      if (iteration > 20) {
        break; // Prevent infinite loop
      }
    }

    if (required_multiplier && required_multiplier.gt(new Decimal(1.001))) {
      // Calculate additional income needed (matching Python logic)
      const original_income =
        this.projection_df[0]?.total_income || new Decimal(0);
      const additional_income = original_income.mul(
        required_multiplier.sub(new Decimal(1.0))
      );

      return {
        type: 'increase_income',
        params: {
          fire_age: this.profile.expected_fire_age,
          percentage: required_multiplier.sub(1).mul(100).toNumber(),
          amount: additional_income.toNumber(),
        },
        is_achievable: true,
        monte_carlo_success_rate: undefined, // Match Python's null behavior
      };
    }

    // If no feasible income increase found within reasonable limits, return null like Python
    return null;
  }

  /**
   * Find required expense reduction using binary search
   */
  private _find_required_expense_reduction(): SimpleRecommendation | null {
    // Binary search for the minimum expense reduction (matching Python logic)
    let low = new Decimal(0.0); // 0% reduction
    let high = new Decimal(0.8); // 80% reduction
    const epsilon = new Decimal(0.001); // Precision for binary search
    let optimal_reduction: Decimal | null = null;

    while (high.sub(low).gt(epsilon)) {
      const mid = low.add(high).div(2);
      const reduction_factor = new Decimal(1.0).sub(mid); // Convert reduction rate to multiplier

      // Create modified projection with reduced expenses
      const modified_projection =
        this._apply_expense_multiplier(reduction_factor);

      // Test this reduction
      const modified_input = createEngineInput(
        this.profile,
        modified_projection,
        this.income_items
      );

      const engine = new FIREEngine(modified_input);
      const result = engine.calculate();

      if (result.is_fire_achievable) {
        optimal_reduction = mid;
        high = mid;
      } else {
        low = mid;
      }
    }

    if (optimal_reduction && optimal_reduction.gt(new Decimal(0.001))) {
      // Calculate annual savings needed (matching Python logic)
      // Find first non-zero expense or use average if all are problematic
      let original_expense = new Decimal(0);

      // Try to find first non-zero expense
      for (const row of this.projection_df) {
        if (row.total_expense && row.total_expense.gt(0)) {
          original_expense = row.total_expense;
          break;
        }
      }

      // If still zero, calculate average expense from all non-zero rows
      if (original_expense.eq(0)) {
        const nonZeroExpenses = this.projection_df
          .map(row => row.total_expense)
          .filter(exp => exp && exp.gt(0));

        if (nonZeroExpenses.length > 0) {
          const sum = nonZeroExpenses.reduce(
            (acc, exp) => acc.add(exp),
            new Decimal(0)
          );
          original_expense = sum.div(nonZeroExpenses.length);
        }
      }

      const annual_savings = original_expense.mul(optimal_reduction);

      return {
        type: 'reduce_expenses',
        params: {
          fire_age: this.profile.expected_fire_age,
          percentage: optimal_reduction.mul(100).toNumber(),
          amount: annual_savings.toNumber(),
        },
        is_achievable: true,
        monte_carlo_success_rate: undefined, // Match Python's null behavior
      };
    }

    // If no feasible expense reduction found within reasonable limits, return null like Python
    return null;
  }

  // =============================================================================
  // Helper Methods for Projection Manipulation
  // =============================================================================

  /**
   * Extend work income to a later FIRE age
   * Direct TypeScript port of Python's _extend_work_income_to_age logic
   */
  private _extend_work_income_to_age(
    target_fire_age: number
  ): AnnualFinancialProjection[] {
    if (!this.income_items || this.income_items.length === 0) {
      throw new Error('Income items required for income extension');
    }

    // Create deep copy to avoid modifying original data
    const extended_projection = this.detailed_projection_df.map(row => ({
      ...row,
    }));
    const current_fire_age = this.profile.expected_fire_age;

    // If target age is not later than current, return unchanged
    if (target_fire_age <= current_fire_age) {
      return extended_projection;
    }

    // Find income items that end at current FIRE age (work income)
    const work_income_items = this.income_items.filter(
      item => item.end_age && item.end_age === current_fire_age
    );

    if (work_income_items.length === 0) {
      // No work income to extend
      return extended_projection;
    }

    // For each work income item, extend it to target age
    for (const item of work_income_items) {
      // Calculate the growth pattern for extending this income
      for (let age = item.end_age + 1; age <= target_fire_age; age++) {
        // Find the row for this age
        const row_index = extended_projection.findIndex(row => row.age === age);

        if (row_index !== -1) {
          // Calculate extended income value with proper growth
          const years_since_start = new Decimal(age - item.start_age);
          const growth_rate_decimal = new Decimal(item.annual_growth_rate).div(
            100
          );
          const growth_factor = new Decimal(1)
            .add(growth_rate_decimal)
            .pow(years_since_start);
          const extended_income =
            item.after_tax_amount_per_period.mul(growth_factor);

          // Set this income directly (not cumulative, since this age shouldn't have work income yet)
          extended_projection[row_index].total_income = extended_income;
        }
      }
    }

    return extended_projection;
  }

  /**
   * Truncate work income to an earlier FIRE age
   */
  private _truncate_work_income_to_age(
    target_fire_age: number
  ): AnnualFinancialProjection[] {
    // Create deep copy to avoid modifying original data
    const modified_projection = this.detailed_projection_df.map(row => ({
      ...row,
    }));
    const current_age = getCurrentAgeAsOf(
      this.profile.birth_year,
      this.profile.as_of_year ?? new Date().getFullYear()
    );

    // Zero out income for years after target FIRE age
    for (let i = 0; i < modified_projection.length; i++) {
      const age = current_age + i;

      if (age > target_fire_age) {
        modified_projection[i].total_income = new Decimal(0);
      }
    }

    return modified_projection;
  }

  /**
   * Apply income multiplier to all income sources
   */
  private _apply_income_multiplier(
    multiplier: Decimal
  ): AnnualFinancialProjection[] {
    return this.projection_df.map(row => ({
      ...row,
      total_income: row.total_income.mul(multiplier),
      net_cash_flow: row.total_income.mul(multiplier).sub(row.total_expense),
    }));
  }

  /**
   * Apply expense multiplier to all expenses
   */
  private _apply_expense_multiplier(
    multiplier: Decimal
  ): AnnualFinancialProjection[] {
    return this.projection_df.map(row => ({
      ...row,
      total_expense: row.total_expense.mul(multiplier),
      net_cash_flow: row.total_income.sub(row.total_expense.mul(multiplier)),
    }));
  }

  /**
   * Create annual summary from detailed DataFrame
   * Handles the case where detailed projection might be the same as annual
   */
  private _create_annual_summary_from_detailed_df(
    detailed_df: AnnualFinancialProjection[]
  ): AnnualFinancialProjection[] {
    // For TypeScript implementation, we assume detailed_df is already in the right format
    // In Python this involves complex pandas grouping, but our data is already annual
    return detailed_df.map(row => ({
      ...row,
      net_cash_flow: row.total_income.sub(row.total_expense),
    }));
  }
}

// =============================================================================
// Factory Functions
// =============================================================================

/**
 * Create a SimpleRecommendation
 */
export function createSimpleRecommendation(
  data: Partial<SimpleRecommendation>
): SimpleRecommendation {
  return {
    type: data.type || '',
    params: data.params || {},
    is_achievable: data.is_achievable ?? true,
    monte_carlo_success_rate: data.monte_carlo_success_rate,
  };
}

/**
 * Create advisor instance - compatibility function for tests
 */
export function createAdvisor(
  engine_input: EngineInput,
  language?: string
): FIREAdvisor {
  return new FIREAdvisor(engine_input, language);
}
